# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1202804149 18000
#      Tue Feb 12 03:15:49 2008 -0500
# Node ID 360b4f7684fdc62ab6647aec762235755f962e43
# Parent  85be2610d6e328b3c15f38d64eb22574fc73535b
fix for loop iteration limit bug with ranges

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,13 @@
+2008-02-12  John W. Eaton  <jwe@octave.org>
+
+	* pt-loop.cc (tree_simple_for_command::eval): Compute range
+	element with multiplication.
+
 2008-02-11  John W. Eaton  <jwe@octave.org>
 
 	* Makefile.in ($(MAKEDEPS)): Skip dependencies if omit_deps is defined.
 
 2008-02-09  John W. Eaton  <jwe@octave.org>
 
 	* pr-output.cc (set_range_format): Eliminate sign arg.
 	Change all callers.
diff --git a/src/pt-loop.cc b/src/pt-loop.cc
--- a/src/pt-loop.cc
+++ b/src/pt-loop.cc
@@ -352,23 +352,31 @@ tree_simple_for_command::eval (void)
     if (rhs.is_range ())
       {
 	Range rng = rhs.range_value ();
 
 	octave_idx_type steps = rng.nelem ();
 	double b = rng.base ();
 	double increment = rng.inc ();
 	bool quit = false;
-	double tmp_val = b;
 
-	for (octave_idx_type i = 0; i < steps; i++, tmp_val += increment)
+	for (octave_idx_type i = 0; i < steps; i++)
 	  {
 	    MAYBE_DO_BREAKPOINT;
 
-	    octave_value val (tmp_val);
+	    // Use multiplication here rather than declaring a
+	    // temporary variable outside the loop and using
+	    //
+	    //   tmp_val += increment
+	    //
+	    // to avoid problems with limited precision.  Also, this
+	    // is consistent with the way Range::matrix_value is
+	    // implemented.
+
+	    octave_value val (b + i * increment);
 
 	    do_for_loop_once (ult, val, quit);
 
 	    if (quit)
 	      break;
 	  }
       }
     else if (rhs.is_scalar_type ())
